crates/kitsune_p2p/kitsune_p2p/src/spawn/actor/space.rs

```diff
@@ -86,85 +83,151 @@ impl gossip::GossipEventHandler for Space {
         &mut self,
     ) -> gossip::GossipEventHandlerResult<Vec<Arc<KitsuneAgent>>> {
         // while full-sync this is just a clone of list_by_basis
-        let res = self.agents.keys().cloned().collect();
-        Ok(async move { Ok(res) }.boxed().into())
+        let all_agents = self
+            .agents
+            .keys()
+            .cloned()
+            .map(|agent| {
+                self.evt_sender
+                    .query_agent_info_signed(QueryAgentInfoSignedEvt {
+                        space: self.space.clone(),
+                        agent,
+                    })
+            })
+            .collect::<Vec<_>>();
+        Ok(async move {
+            let res = futures::stream::iter(all_agents)
+                .buffer_unordered(10)
+                .map_ok(|res| {
+                    res.into_iter()
+                        .map(|ai| Arc::new(ai.into_agent()))
+                        .collect::<Vec<_>>()
+                })
+                .try_collect::<Vec<_>>()
+                .await?
+                .into_iter()
+                .flatten()
+                .collect();
+            Ok(res)
+        }
+        .boxed()
+        .into())
     }
 
     fn handle_req_op_hashes(
         &mut self,
-        _from_agent: Arc<KitsuneAgent>,
-        to_agent: Arc<KitsuneAgent>,
-        dht_arc: kitsune_p2p_types::dht_arc::DhtArc,
-        since_utc_epoch_s: i64,
-        until_utc_epoch_s: i64,
+        input: ReqOpHashesEvt,
     ) -> gossip::GossipEventHandlerResult<OpHashesAgentHashes> {
-        // while full-sync just redirecting to self...
-        // but eventually some of these will be outgoing remote requests
-        let fut = self
-            .evt_sender
-            .fetch_op_hashes_for_constraints(FetchOpHashesForConstraintsEvt {
-                space: self.space.clone(),
-                agent: to_agent.clone(),
+        if self.agents.contains_key(&input.to_agent) {
```

Comment:
should we rename this to something like `self.local_joined_agents` to add clarity?